src/mscorlib/src/System/Globalization/EncodingTable.Unix.cs

```diff
@@ -1,101 +1,155 @@
+            {
+                return codePage;
             }
+
+            // The encoding name is not valid.
+            throw new ArgumentException(
+                string.Format(
+                    CultureInfo.CurrentCulture,
+                    Environment.GetResourceString("Argument_EncodingNotSupported"), name), "name");
         }
-    
-        [System.Security.SecuritySafeCritical]  // auto-generated
-        unsafe internal static CodePageDataItem GetCodePageDataItem(int codepage) {
-            // TODO: Implement this fully.
+
+        internal static CodePageDataItem GetCodePageDataItem(int codepage)
+        {
             switch (codepage)
             {
-                case CodePageUtf7:
-                    return new CodePageDataItem("utf-7", CodePageUtf7, "utf-7", "utf-7", 771);
+                case 1200: return s_encodingDataTableItems[0];
+                case 1201: return s_encodingDataTableItems[1];
+                case 12000: return s_encodingDataTableItems[2];
+                case 12001: return s_encodingDataTableItems[3];
+                case 20127: return s_encodingDataTableItems[4];
+                case 28591: return s_encodingDataTableItems[5];
+                case 65000: return s_encodingDataTableItems[6];
+                case 65001: return s_encodingDataTableItems[7];
+            }
 
-                case CodePageUtf8:
-                    return new CodePageDataItem("utf-8", CodePageUtf8, "utf-8", "utf-8", 771);
+            return null;
+        }
 
-                case CodePageUtf16:
-                    return new CodePageDataItem("utf-16", CodePageUtf16, "utf-16", "utf-16", 771);
+        // PAL ends here.
 
-                case CodePageUtf16BE:
-                    return new CodePageDataItem("utf-16BE", CodePageUtf16BE, "utf-16BE", "utf-16BE", 771);
+#if DEBUG
+        static EncodingTable()
+        {
+            Contract.Assert(
+                s_encodingDataTable.Count == EncodingTableCapacity,
+                string.Format(CultureInfo.InvariantCulture,
+                    "EncodingTable s_encodingDataTable's initial capacity (EncodingTableCapacity) is incorrect.{0}Expected (s_encodingDataTable.Count): {1}, Actual (EncodingTableCapacity): {2}",
+                    Environment.NewLine,
+                    s_encodingDataTable.Count,
+                    EncodingTableCapacity));
+        }
+#endif
 
-                case CodePageUtf32:
-                    return new CodePageDataItem("utf-32", CodePageUtf32, "utf-32", "utf-32", 771);
+        // NOTE: the following two lists were taken from ~\src\classlibnative\nls\encodingdata.cpp
+        // and should be kept in sync with those lists
 
-                default:
-                    return new CodePageDataItem("utf-8", CodePageUtf8, "utf-8", "utf-8", 771);
-            }
-        }
+        private const int EncodingTableCapacity = 42;
+        private readonly static Dictionary<string, ushort> s_encodingDataTable =
```

Comment:
This table is just a mapping of encoding names to code pages.  It is only needed in GetCodePageFromName, which takes the name and returns the code page.
GetCodePageDataItem and GetEncodings don't start with a string name. GetCodePageDataItem already has the code page, so it jumps straight to s_encodingDataTableItems table. GetEncodings returns all the encodings, so it doesn't need the map of the name to the code page.